packages/react-reconciler/src/ReactFiberNewContext.old.js

```diff
@@ -210,43 +218,53 @@ export function propagateContextChange<T>(
-
-            // Inlined `enqueueUpdate` to remove interleaved update check
-            const updateQueue = fiber.updateQueue;
-            if (updateQueue === null) {
-              // Only occurs if the fiber has been unmounted.
-            } else {
-              const sharedQueue: SharedQueue<any> = (updateQueue: any).shared;
-              const pending = sharedQueue.pending;
-              if (pending === null) {
-                // This is the first update. Create a circular list.
-                update.next = update;
+          if (enableLazyContextPropagation) {
+            // In the lazy implemenation, don't mark a dirty flag on the
+            // dependency itself. Not all changes are propagated, so we can't
+            // rely on the propagation function alone to determine whether
+            // something has changed; the consumer will check. In the future,
+            // we could add back a dirty flag as an optimization to avoid
+            // double checking, but until we have selectors it's not really
+            // worth the trouble.
+          } else {
+            if (fiber.tag === ClassComponent) {
+              // Schedule a force update on the work-in-progress.
+              const lane = pickArbitraryLane(renderLanes);
+              const update = createUpdate(NoTimestamp, lane);
+              update.tag = ForceUpdate;
+              // TODO: Because we don't have a work-in-progress, this will add the
+              // update to the current fiber, too, which means it will persist even if
+              // this render is thrown away. Since it's a race condition, not sure it's
+              // worth fixing.
+
+              // Inlined `enqueueUpdate` to remove interleaved update check
+              const updateQueue = fiber.updateQueue;
+              if (updateQueue === null) {
+                // Only occurs if the fiber has been unmounted.
               } else {
-                update.next = pending.next;
-                pending.next = update;
+                const sharedQueue: SharedQueue<any> = (updateQueue: any).shared;
+                const pending = sharedQueue.pending;
+                if (pending === null) {
+                  // This is the first update. Create a circular list.
+                  update.next = update;
+                } else {
+                  update.next = pending.next;
+                  pending.next = update;
+                }
+                sharedQueue.pending = update;
               }
-              sharedQueue.pending = update;
             }
+            // Mark the updated lanes on the list, too.
+            list.lanes = mergeLanes(list.lanes, renderLanes);
           }
+
           fiber.lanes = mergeLanes(fiber.lanes, renderLanes);
           const alternate = fiber.alternate;
           if (alternate !== null) {
             alternate.lanes = mergeLanes(alternate.lanes, renderLanes);
           }
           scheduleWorkOnParentPath(fiber.return, renderLanes);
 
```

Comment:
The problem with not doing the full traversal is that we can't set `DidPropagateContext`. We'd need to replace that with something else to avoid propagating the same tree multiple times.